presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/SpatialPartitioningStateFactory.java

```diff
@@ -0,0 +1,212 @@
+{
+    @Override
+    public SpatialPartitioningState createSingleState()
+    {
+        return new SingleSpatialPartitioningState();
+    }
+
+    @Override
+    public Class getSingleStateClass()
+    {
+        return SpatialPartitioningState.class;
+    }
+
+    @Override
+    public SpatialPartitioningState createGroupedState()
+    {
+        return new GroupedSpatialPartitioningState();
+    }
+
+    @Override
+    public Class getGroupedStateClass()
+    {
+        return GroupedSpatialPartitioningState.class;
+    }
+
+    public static final class GroupedSpatialPartitioningState
+            implements GroupedAccumulatorState, SpatialPartitioningState
+    {
+        private static final int INSTANCE_SIZE = ClassLayout.parseClass(GroupedSpatialPartitioningState.class).instanceSize();
+        private static final int ENVELOPE_SIZE = toIntExact(new Envelope(1, 2, 3, 4).estimateMemorySize());
+
+        private long groupId;
+        private IntBigArray partitionCounts = new IntBigArray();
+        private IntBigArray counts = new IntBigArray();
+        private ObjectBigArray<Rectangle> envelopes = new ObjectBigArray<>();
+        private ObjectBigArray<List<Rectangle>> samples = new ObjectBigArray<>();
+        private int envelopeCount;
+        private int samplesCount;
+
+        @Override
+        public int getPartitionCount()
+        {
+            return partitionCounts.get(groupId);
+        }
+
+        @Override
+        public void setPartitionCount(int partitionCount)
+        {
+            this.partitionCounts.set(groupId, partitionCount);
+        }
+
+        @Override
+        public int getCount()
+        {
+            return counts.get(groupId);
+        }
+
+        @Override
+        public void setCount(int count)
+        {
+            counts.set(groupId, count);
+        }
+
+        @Override
+        public Rectangle getExtent()
+        {
+            return envelopes.get(groupId);
+        }
+
+        @Override
+        public void setExtent(Rectangle envelope)
+        {
+            if (envelopes.get(groupId) == null) {
+                envelopeCount++;
+            }
+            envelopes.set(groupId, envelope);
+        }
+
+        @Override
+        public List<Rectangle> getSamples()
+        {
+            return samples.get(groupId);
+        }
+
+        @Override
+        public void setSamples(List<Rectangle> samples)
+        {
+            List<Rectangle> currentSamples = this.samples.get(groupId);
+            if (currentSamples != null) {
+                samplesCount -= currentSamples.size();
+            }
+            samplesCount += samples.size();
+            this.samples.set(groupId, samples);
+        }
+
+        @Override
+        public long getEstimatedSize()
+        {
+            return INSTANCE_SIZE + partitionCounts.sizeOf() + counts.sizeOf() + envelopes.sizeOf() + samples.sizeOf() + ENVELOPE_SIZE * (envelopeCount + samplesCount);
```

Comment:
Create `SingleState#estimateMemorySize()` method. Rewrite it to `INSTANCE_SIZE + states.sizeOf() + sum of sizes of single state`.

Reply:
@arhimondr `ObjectBigArray` doesn't support iterating over the elements, hence, I can't compute `sum of sizes of single state`